Will now be fitting linear regression with multiple features and completing all the math/notes with numpy

In [1]:
import numpy as np

In [52]:
# Matrix of examples along with vector corresponding to outputs
X = np.array([
    [1.2, 3.4, 5.1, 2.2],
    [2.3, 0.4, 1.1, 3.3],
    [0.9, 4.5, 2.2, 1.1],
    [3.3, 2.2, 0.5, 4.0],
    [1.1, 1.0, 4.7, 2.8],
    [2.5, 3.8, 1.4, 0.9],
    [4.1, 0.9, 0.8, 2.6],
    [3.7, 2.1, 3.3, 1.5],
    [0.6, 4.9, 2.7, 3.9],
    [1.9, 1.8, 4.6, 2.4],
    [2.4, 3.3, 1.6, 1.2],
    [3.8, 2.7, 0.9, 3.1],
    [0.7, 4.2, 3.0, 2.0],
    [1.5, 1.3, 4.1, 3.6],
    [2.6, 3.1, 1.8, 0.7],
    [3.9, 2.9, 0.6, 1.9],
    [0.8, 4.4, 2.5, 2.7],
    [1.7, 1.6, 4.3, 1.8],
    [2.9, 3.6, 1.3, 0.6],
    [3.4, 2.4, 0.7, 3.4]
])

y = np.array([
    10.2, 5.3, 8.7, 6.1, 9.5,
    7.2, 4.8, 6.9, 9.1, 8.6,
    7.4, 6.3, 8.4, 9.0, 7.0,
    6.5, 8.2, 8.9, 7.1, 6.7
])

In [18]:
# Returns the cost of the given LR model when passed w and b
def cost(X, y, W, b):
    m = X.shape[0]
    y_hat = np.dot(X, W) + b
    error = y_hat - y

    total_cost = (np.sum(error ** 2)) / (2 * m)
    return total_cost

W = np.array([0, 0, 0, 0])
b = 0

print(f"Cost is {cost(X, y, W, b)}")


Cost is 29.838750000000005


With default values of 0 for the weights and intercept we see a cost of 29.8, after gradient descent lets see how this can be improved

In [80]:
# Immediately going to vectorized gradient descent this time - without feature scaling
def v_gradientDescent(X, y, learningRate, iterations):
    m = X.shape[0]
    W = np.array([0, 0, 0, 0])
    b = 0
    dw = 0
    db = 0

    for i in range(iterations):
        y_hat = np.dot(X, W) + b
        error = y_hat - y

        dw = (np.dot(X.T, error)) / m
        db = (np.sum(error)) / m

        W = W - (learningRate * dw)
        b = b - (learningRate * db)

        if (i % 20 == 0):
            print(f"Cost is currently {cost(X, y, W, b)}")

    return W, b

w, b = v_gradientDescent(X, y, 0.1, 4000)



Cost is currently 64.13325610110564
Cost is currently 363146561.1037983
Cost is currently 2069093840421378.5
Cost is currently 1.1789039967585798e+22
Cost is currently 6.717020786695265e+28
Cost is currently 3.827145244477089e+35
Cost is currently 2.180585885834343e+42
Cost is currently 1.2424286254517699e+49
Cost is currently 7.07896395812607e+55
Cost is currently 4.0333689753990114e+62
Cost is currently 2.2980856221250933e+69
Cost is currently 1.3093762457216321e+76
Cost is currently 7.460410249095376e+82
Cost is currently 4.2507049648004683e+89
Cost is currently 2.421916770591843e+96
Cost is currently 1.3799313036889065e+103
Cost is currently 7.862410575055517e+109
Cost is currently 4.479751990949914e+116
Cost is currently 2.5524204960865976e+123
Cost is currently 1.4542881842576102e+130
Cost is currently 8.286072479491396e+136
Cost is currently 4.721141096971365e+143
Cost is currently 2.6899563469519783e+150
Cost is currently 1.532651746661219e+157
Cost is currently 8.7325631852925

C:\Users\darth\AppData\Local\Temp\ipykernel_18708\2650675279.py:7: RuntimeWarning: overflow encountered in square
  total_cost = (np.sum(error ** 2)) / (2 * m)
C:\Users\darth\AppData\Local\Temp\ipykernel_18708\2103423673.py:13: RuntimeWarning: overflow encountered in dot
  dw = (np.dot(X.T, error)) / m
C:\Users\darth\AppData\Local\Temp\ipykernel_18708\2103423673.py:16: RuntimeWarning: invalid value encountered in subtract
  W = W - (learningRate * dw)


With all of our values already being closely correlated in the input data, it is unlikely that regularization will make significant difference in the outcome but we will test that along with a different learning rate. As we can see 0.1 causes divergence and our cost is not converging at 0 but rather fluctuating

In [78]:
w, b = v_gradientDescent(X, y, 0.08, 4000)

Cost is currently 28.563105960707606
Cost is currently 12.893217294884915
Cost is currently 5.921779782993663
Cost is currently 2.777958372558533
Cost is currently 1.3589064471512906
Cost is currently 0.7172536010764751
Cost is currently 0.4260401498789066
Cost is currently 0.2928367797806179
Cost is currently 0.23091468759544007
Cost is currently 0.20118354324737808
Cost is currently 0.18602339333335044
Cost is currently 0.17749089716436064
Cost is currently 0.17200396432284215
Cost is currently 0.1679461586228137
Cost is currently 0.16458701869664077
Cost is currently 0.16159545229530672
Cost is currently 0.15882042448883726
Cost is currently 0.1561921132766871
Cost is currently 0.15367735902145852
Cost is currently 0.15125958377044713
Cost is currently 0.14892973904127216
Cost is currently 0.14668222412153148
Cost is currently 0.14451304460804257
Cost is currently 0.14241898066940886
Cost is currently 0.14039721044537054
Cost is currently 0.13844513863285965
Cost is currently 0.1365

With an adjusted learning rate of 0.8, we can now see the data converging closer to 0. Below gradient descent will be re-implemented but with feature scaling for the sake of trying to get an even lower cost

In [ ]:
# Feature scaling our input matrix X - ZScore Normalization
X_means = np.zeros(X.shape[1])
X_stds = np.zeros(X.shape[1])

for i in range(X.shape[1]):
    X_means[i] = np.mean(X[:, i])
    X_stds[i] = np.std(X[:, i])

X_scaled = (X - X_means) / X_stds
print(X_scaled)

[[-0.93110806  0.53051432  1.83385315 -0.08159227]
 [ 0.0305998  -1.82732712 -0.84330473  0.97430772]
 [-1.19339202  1.39505619 -0.10708632 -1.13749226]
 [ 0.90487966 -0.41262225 -1.24487842  1.64624408]
 [-1.01853605 -1.35575883  1.56613736  0.49435318]
 [ 0.20545577  0.84489318 -0.64251789 -1.32947408]
 [ 1.60430356 -1.43435354 -1.04409158  0.30237136]
 [ 1.25459161 -0.49121697  0.6291321  -0.75352863]
 [-1.45567598  1.70943504  0.22755842  1.55025317]
 [-0.31911215 -0.72700111  1.49920842  0.11038954]
 [ 0.11802778  0.45191961 -0.50866    -1.04150135]
 [ 1.3420196  -0.01964868 -0.97716263  0.7823259 ]
 [-1.368248    1.15927204  0.42834526 -0.27357409]
 [-0.6688241  -1.11997468  1.16456368  1.26228044]
 [ 0.29288376  0.29473018 -0.3748021  -1.52145589]
 [ 1.42944759  0.13754075 -1.17794947 -0.369565  ]
 [-1.28082001  1.31646147  0.09370053  0.39836227]
 [-0.49396813 -0.88419054  1.29842157 -0.4655559 ]
 [ 0.55516772  0.68770375 -0.70944684 -1.6174468 ]
 [ 0.99230765 -0.25543282 -1.11

Now these scaled X values hover around 0 and we can see how this impacts gradient descent

In [82]:
w, b = v_gradientDescent(X_scaled, y, 0.01, 1000)

Cost is currently 29.234955272815984
Cost is currently 19.453554889450924
Cost is currently 12.9803531321376
Cost is currently 8.684770296207436
Cost is currently 5.828265892276974
Cost is currently 3.92565137277942
Cost is currently 2.6567773667931243
Cost is currently 1.8096824446297846
Cost is currently 1.2436726241515563
Cost is currently 0.8651771826076221
Cost is currently 0.6118736556874358
Cost is currently 0.4422060681358465
Cost is currently 0.3284419015315646
Cost is currently 0.252061843813704
Cost is currently 0.20069286258802332
Cost is currently 0.16606510864379226
Cost is currently 0.14264943253308976
Cost is currently 0.12674806362737975
Cost is currently 0.11588736855011339
Cost is currently 0.10841215571569725
Cost is currently 0.1032145404557427
Cost is currently 0.0995526934786113
Cost is currently 0.0969296519197925
Cost is currently 0.095012277287499
Cost is currently 0.0935770540318834
Cost is currently 0.09247383559473657
Cost is currently 0.09160159285794947
C

Where this learning rate was too large for the non-feature scaled matrix, it actually begins to convege with the scaling.

In [ ]:
w, b = v_gradientDescent(X_scaled, y, 0.08, 500)

Cost is currently 25.184317234044496
Cost is currently 0.9768250824346196
Cost is currently 0.12772738507423903
Cost is currently 0.09260685412899422
Cost is currently 0.08840445709687596
Cost is currently 0.08649092913843563
Cost is currently 0.08536868364113018
Cost is currently 0.08469877057661945
Cost is currently 0.08429844591059976
Cost is currently 0.08405920579476235
Cost is currently 0.08391623170920856
Cost is currently 0.08383078787031242
Cost is currently 0.08377972512224326
Cost is currently 0.08374920912159896
Cost is currently 0.0837309722203236
Cost is currently 0.08372007352583209
Cost is currently 0.08371356027339803
Cost is currently 0.08370966783863872
Cost is currently 0.08370734165083808
Cost is currently 0.08370595147989912
Cost is currently 0.08370512068917832
Cost is currently 0.08370462419396159
Cost is currently 0.08370432747964567
Cost is currently 0.08370415015792694
Cost is currently 0.0837040441873342


Using our learning rate of 0.8 again, convergence was able to be located much faster than previously with only around 500 iterations needed to see convergence